pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -846,61 +847,155 @@ func (m *kubeGenericRuntimeManager) purgeInitContainers(ctx context.Context, pod
+		if podHasInitialized && !types.IsSidecarContainer(container) {
+			// after initialization, only sidecar containers need to be kept running
+			continue
 		}
-	}
 
-	// There are no failed containers now.
-	for i := len(pod.Spec.InitContainers) - 1; i >= 0; i-- {
-		container := &pod.Spec.InitContainers[i]
 		status := podStatus.FindContainerStatusByName(container.Name)
+		klog.V(4).InfoS("Computing init container action", "pod", klog.KObj(pod), "container", container.Name, "status", status)
+		// If the container is not found, it means it has not been started yet.
 		if status == nil {
-			continue
+			if podHasInitialized || previousContainerInitialized {
+				// For the new init container,
+				// - if the pod has been initialized before, we should restart the
+				// sidecar container here as it should restart always.
+				// - if the previous init container has been initialized, it is time to
+				// start the next one.
+				changes.InitContainersToStart = append(changes.InitContainersToStart, i)
+			}
+			// There is an init container that has not been started yet.
+			// We can stop here.
+			return false
 		}
 
-		// container is still running, return not done.
-		if status.State == kubecontainer.ContainerStateRunning {
-			return nil, nil, false
-		}
+		previousContainerInitialized = false
+		switch status.State {
+		case kubecontainer.ContainerStateCreated:
+			switch {
+			case types.IsSidecarContainer(container):
+			default: // init container
+				return false
+			}
+
+		case kubecontainer.ContainerStateRunning:
+			switch {
+			case types.IsSidecarContainer(container):
+				if container.StartupProbe != nil {
+					startup, found := m.startupManager.Get(status.ID)
+					if !found {
+						// If the startup probe has not been run, wait for it.
+						continue
+					}
+					if startup != proberesults.Success {
+						if startup == proberesults.Failure {
+							// If the sidecar container failed the startup probe, restart it.
+							changes.ContainersToKill[status.ID] = containerToKillInfo{
+								name:      container.Name,
+								container: container,
+								message:   fmt.Sprintf("Container %s failed startup probe", container.Name),
+								reason:    reasonStartupProbe,
+							}
+							changes.InitContainersToStart = append(changes.InitContainersToStart, i)
+						}
+						continue
+					}
+				}
+				previousContainerInitialized = true
+			default: // init container
+				return false
+			}
 
-		if status.State == kubecontainer.ContainerStateExited {
-			// all init containers successful
-			if i == (len(pod.Spec.InitContainers) - 1) {
-				return nil, nil, true
+		case kubecontainer.ContainerStateExited:
+			switch {
+			case types.IsSidecarContainer(container):
+				changes.InitContainersToStart = append(changes.InitContainersToStart, i)
+			default: // init container
+				if isInitContainerFailed(status) {
+					if !restartOnFailure {
+						changes.KillPod = true
+						changes.InitContainersToStart = nil
```

Comment:
can we add a few comments to explain the logic here to the top of the `kubecontainer.ContainerStateExited:` to summarize? 

The logic is:

(1) If the init container failed and restart policy is never - then the pod is terminal.
(2) Otherwise, If the init container failed, restart that init container